### PR TITLE
Update struct_expr grammar for field init shorthand.

### PR DESCRIPTION
**Pull request by theduke**
Part of the work for #38830 .

r? @steveklabnik

### CODE REVIEW

File: src/doc/grammar.md [hunk 1]
```diff
@@ -510,8 +510,8 @@ unit_expr : "()" ;
 ### Structure expressions
 
 ```antlr
-struct_expr : expr_path '{' ident ':' expr
-                      [ ',' ident ':' expr ] *
+struct_expr : expr_path '{' ident [ ':' expr ] ?
+                      [ ',' [ ident [ ':' expr ] ? ] *
```
Comment (nagisa) [inline src/doc/grammar.md]: This is incorrect.

Comment (theduke) [inline src/doc/grammar.md]: Why ?

Comment (theduke) [inline src/doc/grammar.md]: Because of the ambiguity?

```markdown
  ident | ident ':' expr
```

Comment (nagisa) [inline src/doc/grammar.md]: Braces are not balanced (and there’s the same mistake in the cleaner version as well). Extracting into `struct_expr_field_init` sounds like a good idea regardless. Makes it easier to read.

Comment (theduke) [inline src/doc/grammar.md]: Ah yeah, I'm toying with alacritty and it messes up my brackets in vim currenty, but I obvisouly should have noticed that, my bad.

Should be fixed now.

**Comment by nagisa** [comment]
This LGTM.

**Comment by steveklabnik** [comment]
@bors: r+ rollup

sorry for the tardiness here!

**Pull request merged**